#462103 [C#テキスト]11章演習問題

kurosu-being opened the pull request:
第11章 XMLファイルの操作

主にLINQ to XMLを使ったXMLファイルの操作を学習しました。
また、XMLファイルのフォーマットについての学習も行いました。

kurosu-being committed:
#462103 [C#テキスト]11章演習問題

Practice11-1-1のファイルを作成しました。

kurosu-being committed:
#462103 [C#テキスト]11章演習問題

Practice11-1-1を実装しました。

kurosu-being committed:
#462103 [C#テキスト]11章演習問題

Practice11-1-1において、不要なusingを削除しました。
Practice11-1-2を実装しました。

kurosu-being committed:
#462103 [C#テキスト]11章演習問題

Practice11-1-2において、ToList()を削除し、wSortkanjiNamesを削除しました。
Practice11-1-3のファイルを作成しました。

kurosu-being committed:
#462103 [C#テキスト]11章演習問題

Practice11-1-1の.NetFrameWorkを2.6.1→2.7.1に変更しました。
Practice11-1-3を実装しました。

kurosu-being committed:
#462103 [C#テキスト]11章演習問題

Practice11-1-3の不要な変数を削除しました。
Practice11-1-4を実装しました。

kurosu-being committed:
#462103 [C#テキスト]11章演習問題

Practice11-1-4の.フォーマットを修正しました。

kurosu-being committed:
#462103 [C#テキスト]11章演習問題

Practice11-1-1~11-1-4まで、読み込むファイルが存在しなかった時retrunで返すように修正しました。
Practice11-1-4において、出力するファイル名を設定してコード上でファイルを作成するように修正しました。

kurosu-being committed:
#462103 [C#テキスト]11章演習問題

Practice11-2-1のファイルを作成しました。

kurosu-being committed:
#462103 [C#テキスト]11章演習問題

Practice11-2-1において、問題文を記載しました。

kurosu-being committed:
#462103 [C#テキスト]11章演習問題

Practice11-1-1～Practice11-1-4において、最初に表示される指示が明確になるように変更しました。
Practice11-2-1を実装しました。

k-miyahara-being commented on 章末問題11/Practice11-1-1/Program11-1-1.cs:
```diff
@@ -0,0 +1,42 @@
+            var wXdocument = XDocument.Load(wInputFile);
+            foreach (var wGameName in wXdocument.Root.Elements()) {
+                var wXname = wGameName.Element("name");
+                var wXTeamMember = wGameName.Element("teammembers");
```
このあたり、明示的な型を使用しましょう。
ただし変数なくても良いかもです

kurosu-being replied:
ご指摘ありがとうございます。
不要な変数を削除し、明示的な型を使用して修正しました。

k-miyahara-being commented on 章末問題11/Practice11-1-1/Program11-1-1.cs:
```diff
@@ -0,0 +1,42 @@
+                return;
+            }
+            var wXdocument = XDocument.Load(wInputFile);
+            foreach (var wGameName in wXdocument.Root.Elements()) {
```
wGameNameについて変数名を再度考えてみてください。
ループしているのは、「競技名」ではなく「競技」もしくは「球技」ですね。

kurosu-being replied:
ご指摘ありがとうございます。
ループしているものを正しく示すように変数名を変更しました。（wGameName→wGame)

k-miyahara-being commented on 章末問題11/Practice11-1-2/Program11-1-2.cs:
```diff
@@ -0,0 +1,28 @@
+                return;
+            }
+
+            Console.WriteLine("最初にプレーされた年の若い順に漢字の表記名を表示します");
```
同じことを最初に言っているので、このメッセージは冗長かなと思います。

kurosu-being replied:
ご指摘ありがとうございます。
メッセージを削除しました。

k-miyahara-being commented on 章末問題11/Practice11-1-2/Program11-1-2.cs:
```diff
@@ -0,0 +1,28 @@
+                var wXname = wGameName.Element("name");
+                XAttribute wXAttribute = wXname.Attribute("kanji");
+
+                Console.WriteLine($"漢字表記名「{wXAttribute?.Value}」");
```
しっかりnullチェックされてますね！
たしかに、漢字属性がない場合nullとなります。

では、nullまたは空のときは「なし」と表示されるようにしてみてください。
nullなどの異常値を判定するときは、実際そのチェックにかかった時どうするか、まで考えると良いです。

kurosu-being replied:
ありがとうございます！

null、または空の時は「なし」と表示されるように変更いたしました。

k-miyahara-being commented on 章末問題11/Practice11-1-4/Program11-1-4.cs:
```diff
@@ -0,0 +1,38 @@
+
+            var wXdocuments = XDocument.Load(wInputFile);
+            var wOption = new XElement("ballsport");
+            wOption.SetElementValue("name", "サッカー");
+            wOption.SetElementValue("kanji", "蹴球");
```
kanjiは属性ですが、要素として追加されてしまっています。

XEmelemtのコンストラクタのオーバーロードなどをみてみるとよいです。

kurosu-being replied:
ご指摘ありがとうございます。
kanjiが属性として追加されるよう、修正いたしました。

k-miyahara-being commented on 章末問題11/Practice11-1-4/Program11-1-4.cs:
```diff
@@ -0,0 +1,38 @@
+
+            wXdocuments.Root.Add(wRoot);
+
+            Console.WriteLine("新しいXMLファイルの名前を入力してください。");
+
+            var wOutputFileName = Console.ReadLine();
+            var wOutputFile = Path.Combine(Path.GetDirectoryName(wInputFile), wOutputFileName + ".xml");
```
入力したファイル名に拡張子(.xml)が既についているならそのまま、
ついていないなら拡張子を付ける、という風に実装してみてください。

エクスプローラでファイル保存するときは、↑のような動きになっています。

kurosu-being replied:
> 入力したファイル名に拡張子(.xml)が既についているならそのまま、
> ついていないなら拡張子を付ける、という風に実装してみてください。

そのような挙動になるよう機能を追加しました。
エクスプローラーもこのような形になっていることを初めて知りました😲

k-miyahara-being commented on 章末問題11/Practice11-2-1/Program11-2-1.cs:
```diff
@@ -0,0 +1,73 @@
+                // XMLファイルを読み込む
+                var wDocument = XDocument.Load(wInputFile);
+
+                foreach (var wOrdElement in wDocument.Descendants("word")) {
```
wOrdElement 誤字ですかね？
接頭辞wとwordがかぶった？😂

kurosu-being replied:
被りました😭
ご指摘ありがとうございます。修正いたしました。

k-miyahara-being commented on 章末問題11/Practice11-2-1/Program11-2-1.cs:
```diff
@@ -0,0 +1,73 @@
+                    return;
+                }
+
+                if (Path.GetExtension(wInputFile) != "xml") {
+                    Console.WriteLine("XMLファイルではありません。XMLファイルのパスを入力してください.");
+                    return;
```
拡張子には"."が必要です。
`"xml"` → `".xml"`

kurosu-being replied:
ご指摘ありがとうございます。
修正いたしました。

kurosu-being committed:
#462103 [C#テキスト]11章演習問題

Practice11-1-1とPractice11-2-1において、wGameName→wGameに変更しました。

kurosu-being committed:
#462103 [C#テキスト]11章演習問題

Practice11-1-2において、

kurosu-being committed:
#462103 [C#テキスト]11章演習問題

Practice11-1-2において、漢字表記名がnullまたは空の時漢字表記名なしと表示されるよう修正しました。

kurosu-being committed:
#462103 [C#テキスト]11章演習問題

Practice11-1-2において、kanjiが属性として追加されるよう修正しました。

kurosu-being committed:
#462103 [C#テキスト]11章演習問題

Practice11-1-4において、入力したファイル名に拡張子(.xml)が既についているならそのまま、ついていないなら拡張子を付ける機能を追加しました。
Practice11-2-1において、誤字を修正しました。(wOrdElement→wWordElement)

kurosu-being committed:
#462103 [C#テキスト]11章演習問題

Practice11-1-2において、漢字表記名がnullまたは空の時なしと表示されるよう修正しました。
